wallet/proto_wallet.go

```diff
@@ -0,0 +1,353 @@
+package wallet
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"crypto/hmac"
+	"crypto/sha256"
+
+	ec "github.com/bsv-blockchain/go-sdk/primitives/ec"
+	hash "github.com/bsv-blockchain/go-sdk/primitives/hash"
+)
+
+// ProtoWallet is a precursor to a full wallet, capable of performing foundational cryptographic operations.
+// It can derive keys, create signatures, facilitate encryption and HMAC operations.
+// Unlike a full wallet, it doesn't create transactions, manage outputs, interact with the blockchain,
+// or store any data.
+type ProtoWallet struct {
+	// The underlying key deriver
+	keyDeriver *KeyDeriver
+}
+type ProtoWalletArgsType string
+
+const (
+	ProtoWalletArgsTypePrivateKey ProtoWalletArgsType = "privateKey"
+	ProtoWalletArgsTypeKeyDeriver ProtoWalletArgsType = "keyDeriver"
+	ProtoWalletArgsTypeAnyone     ProtoWalletArgsType = "anyone"
+)
+
+type ProtoWalletArgs struct {
+	Type       ProtoWalletArgsType
+	PrivateKey *ec.PrivateKey
+	KeyDeriver *KeyDeriver
+}
+
+// NewProtoWallet creates a new ProtoWallet from a private key or KeyDeriver
+func NewProtoWallet(rootKeyOrKeyDeriver ProtoWalletArgs) (*ProtoWallet, error) {
+	switch rootKeyOrKeyDeriver.Type {
+	case ProtoWalletArgsTypeKeyDeriver:
+		return &ProtoWallet{
+			keyDeriver: NewKeyDeriver(rootKeyOrKeyDeriver.PrivateKey),
+		}, nil
+	case ProtoWalletArgsTypePrivateKey:
+		return &ProtoWallet{
+			keyDeriver: NewKeyDeriver(rootKeyOrKeyDeriver.PrivateKey),
+		}, nil
+	case ProtoWalletArgsTypeAnyone:
+		// Create an "anyone" key deriver as default
+		kd := NewKeyDeriver(nil)
+		return &ProtoWallet{
+			keyDeriver: kd,
+		}, nil
+	}
+	return nil, errors.New("invalid rootKeyOrKeyDeriver")
+}
+
+// GetPublicKey returns the public key for the wallet
+func (p *ProtoWallet) GetPublicKey(ctx context.Context, args GetPublicKeyArgs, _originator string) (*GetPublicKeyResult, error) {
```

Comment:
consider checking `ctx.Done` so we won't burn the resources for calculations if context is already done.